stream-chat-android-core/src/main/java/io/getstream/chat/android/core/utils/Debouncer.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.core.utils
+
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+/**
+ * Utility class for debouncing high frequency events.
+ *
+ * [submit]ting a new piece of work to run within the debounce window
+ * will cancel the previously submitted pending work.
+ */
+public class Debouncer(
+    private val debounceMs: Long,
+    private val scope: CoroutineScope = CoroutineScope(DispatcherProvider.Main),
+) {
+
+    private var job: Job? = null
+
+    /**
+     * Cancels the previous work and launches a new coroutine
+     * containing the new work.
+     */
+    public fun submit(work: () -> Unit) {
+        job?.cancel()
+        job = scope.launch {
+            delay(debounceMs)
+            work()
+        }
+    }
+
+    /**
+     * Cancels the previous work and launches a new coroutine
+     * containing the new suspendable work.
+     */
+    public fun submitSuspendable(work: suspend () -> Unit) {
+        job?.cancel()
+        job = scope.launch {
+            delay(debounceMs)
+            work()
+        }
+    }
```

Comment:
Isn't debouncer behavior a bit weird?
Shouldn't it run the job and reject any new job until the delay time expires?
Maybe the task that this PR tries to fix is not a problem, but imagine we are using this debouncer class in another place to throttling an action that is executed multiple times in a second and we want to limit it to only once per second, this will cause the action to not being executed.

Reply:
For throttling I would prefer to have a different helper class, `Throttler` maybe.

Reply:
You are right